### PR TITLE
Feature clasp take 2

### PR DESCRIPTION
**Pull request by kpoeck**
Tests done with clasp:
* (xlib-demo/demos:demo)
* (clim-demo:demodemo) 

To test for regressions in other lisps, test with sbcl and ccl
* (xlib-demo/demos:demo)

as I put in the code:
define-constant-uneql is needed, since constants with non-eql value have undefined behaviour when redefined. Compiling and Loading and file may redefine the constant, please see  http://www.sbcl.org/manual/#Defining-Constants. This interpretation is shared by sbcl and clasp

### CODE REVIEW

File: clx.lisp [hunk 1]
```diff
@@ -438,7 +438,7 @@
 
 (deftype xatom () '(or string symbol))
 
-(defconstant +predefined-atoms+
+(define-constant-uneql +predefined-atoms+
```
Comment (dkochmanski) [inline clx.lisp]: I don't like this solution. I think that something like below will be less invasive. I'm thinking about patching defconstant for all implementations like this.

```
From 243ca7ea2de52845cb7808852c1b9a8ce3ced6db Mon Sep 17 00:00:00 2001
From: Daniel Kochmanski <daniel@turtleware.eu>
Date: Sun, 6 Jan 2019 13:00:44 +0100
Subject: [PATCH] depdefs: patch defconstant for implementations which insist
 on eql

Most notably clasp and sbcl right now. We may consider making it default for all
implementations.
---
 depdefs.lisp | 8 ++++++++
 package.lisp | 1 +
 2 files changed, 9 insertions(+)

diff --git a/depdefs.lisp b/depdefs.lisp
index 07e1999..84de96b 100644
--- a/depdefs.lisp
+++ b/depdefs.lisp
@@ -35,6 +35,14 @@
 	(lisp:rational x)))
   (deftype rational (&optional l u) `(lisp:rational ,l ,u)))
 
+;;; Fix an issue with implementations which insist that constant values must be
+;;; equal. Redefine defconstant macro to ensure a desired behavior.
+#+(or clasp sbcl)
+(defmacro defconstant (name value &optional doc)
+  `(cl:defconstant ,name
+     (if (boundp ',name) (symbol-value ',name) ,value)
+     ,@(when doc (list doc))))
+
 ;;; CLX-VALUES value1 value2 ... -- Documents the values returned by the function.
 
 #-Genera
diff --git a/package.lisp b/package.lisp
index 6550903..41b2726 100644
--- a/package.lisp
+++ b/package.lisp
@@ -23,6 +23,7 @@
 (defpackage #:xlib
   (:use common-lisp)
   (:size 3000)
+  #+(or sbcl clasp) (:shadow defconstant)
   #+(or kcl ibcl) (:shadow rational)
   #+allegro (:import-from excl without-interrupts)
   #+excl (:import-from excl arglist)
-- 
2.17.1
```

Comment (kpoeck) [reply clx.lisp]: What is the change you require here? (a) I undo all define-constant-uneql and shadow defconstant as you propose or (b) I just undo the define-constant-uneql and wait for you to shadow defconstant  separately?

**Comment by dkochmanski** [comment]
Please undo and add my change unstead.

Dnia 6 stycznia 2019 13:43:47 CET, Karsten Poeck <notifications@github.com> napisał(a):
>kpoeck commented on this pull request.
>
>
>
>> @@ -438,7 +438,7 @@
> 
> (deftype xatom () '(or string symbol))
> 
>-(defconstant +predefined-atoms+
>+(define-constant-uneql +predefined-atoms+
>
>What is the change you require here? (a) I undo all
>define-constant-uneql and shadow defconstant as you propose or (b) I
>just undo the define-constant-uneql and wait for you to shadow
>defconstant  separately?
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/sharplispers/clx/pull/136#discussion_r245507295

-- Wysłane za pomocą K-9 Mail.

File: clx.lisp (hunk 1 above)
Comment (dkochmanski) [reply clx.lisp]: version a) (comment below was sent from mail)

**Comment by kpoeck** [comment]
I hope I have done all what you requested. Tested with clasp, sbcl and ccl

**Comment by dkochmanski** [comment]
lgtm, thank you. normally I'd ask you to squash first and third commit, but I'll simply squash all of them so there is no need to take an action on your side.

nb: in my ideal world you'd test it with ecl too :-)

File: clx.lisp (hunk 1 above)
Comment (Harleqin) [reply clx.lisp]: Alexandria has `define-constant`, which should be used (or copied) here, I think.

Comment (dkochmanski) [reply clx.lisp]: you are right, I don't know why I haven't thought about that. Thank you!

**Pull request merged**